### ISSUE TITLE
API to allow coroutine tail recursion optimization

### ISSUE THREAD

**Issue by h0tk3y** [Kotlin/kotlin-coroutines#67]
Recursive generators mentioned in https://github.com/Kotlin/kotlinx.coroutines/issues/7 could benefit from tail recursion optimization. Shortly, the use case is

```
fun fromTo(from: Int, to: Int) = generate {
    if (from > to) return
    yield(from)
    yieldAll(fromTo(from + 1, to))
}
```

Efficient recursion can be implemented here, but there's no transparent way to implement tailrec optimization inside the library code. 

Possible ways, however, include:
- Using handleResult passing the _tail_ as a result:
  
  ```
  fun fromTo(from: Int, to: Int) = generateWithTail<Int> f@{
     if (from > to) return@f emptySequence()
     yield(from)
     return@f fromTo(from + 1, to)
  }
  ```
  
  This would also benefit from https://github.com/Kotlin/kotlin-coroutines/issues/65 as it allows using `return@f` with no return value in line 2
- Coroutines API that would allow one to find out that a Continuation<Unit> is basically a NOP that will do nothing but immediately finish the coroutine block. Once we know this about continuations we can easily optimize tail recursion.

**Comment by elizarov** [comment]
With the revised coroutine implementation this can be directly written as tailrec without the need for `yieldAll` or any other magic whatsoever:

```
tailrec suspend fun Generator<Int>.fromTo(from: Int, to: Int) {
    if (from > to) return
    yield(from)
    return fromTo(from + 1, to)
}
```

It does not work in 1.1-M04 yet, because non-tail `yield` is not allowed yet, but is planned to be implemented.

**Comment by elizarov** [comment]
*Update*: It is still an issue in Kotlin 1.1.1: https://youtrack.jetbrains.com/issue/KT-16950

**Comment by elizarov** [comment]
It is fixed in 1.1.2.
